### PR TITLE
Move `nullif` and `isnan` to datafusion-functions

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/8045
See also #9100

## Rationale for this change

The high level rationale is described in https://github.com/apache/arrow-datafusion/issues/8045

In this PR I am trying to set up the pattern for how functions can be moved to the `datafusion-functions` crate so we can continue the wholesale movement.

I specifically wanted `core` to unblock the movement of `arrow_cast` -- https://github.com/apache/arrow-datafusion/issues/9143#issuecomment-1941233120 from @brayanjuls 

## What changes are included in this PR?

1. Move `isnan` function to `datafusion-functions` crate, math_expressions
2. Move 'nullif' function to `datafusion-functions` crate, core_expressions

## Are these changes tested?
Yes, covered by existing tests and new CI test

## Are there any user-facing changes?

I don't think so

### CODE REVIEW

File: datafusion/sql/tests/sql_integration.rs [hunk 1]
```diff
@@ -2671,13 +2673,62 @@ fn logical_plan_with_dialect_and_options(
...
+        DataType::Int32,
+    ));
+
     let planner = SqlToRel::new_with_options(&context, options);
     let result = DFParser::parse_sql_with_dialect(sql, dialect);
     let mut ast = result?;
     planner.statement_to_plan(ast.pop_front().unwrap())
 }
 
+fn make_udf(name: &'static str, args: Vec<DataType>, return_type: DataType) -> ScalarUDF {
```
Comment (alamb) [inline datafusion/sql/tests/sql_integration.rs:2688]: The sql integration test needs function definitions, to resolve function names, but doesn't actually invoke them

File: datafusion/functions/src/macros.rs [hunk 2]
```diff
@@ -121,3 +121,42 @@ macro_rules! make_package {
...
+    }};
+}
+
+/// Downcast an argument to a specific array type, returning an internal error
+/// if the cast fails
+///
+/// $ARG: ArrayRef
+/// $NAME: name of the argument (for error messages)
+/// $ARRAY_TYPE: the type of array to cast the argument to
+macro_rules! downcast_arg {
```
Comment (comphead) [inline datafusion/functions/src/macros.rs:152]: I remember to move those downcast macros :)

**Review by comphead — APPROVED** [review]
lgtm, thanks @alamb

**Comment by alamb** [comment]
Thank you @comphead 🙏

**Pull request merged**